### PR TITLE
v2.x asm updates

### PR DESCRIPTION
**Pull request by hjelmn**
This PR brings over most of the assembly changes from master. This includes:

 - ARMv8 support
 - __sync builtin support (this is really desirable since the __builtin atomics are deprecated in gcc 5.x)
 - Allow PGI to use the inline assembly. We work around the bugs for ppc in 16.x so no need to have it disabled.
 - Add checks that builtin atomics support 64-bit values. This fixes a bug on some systems when using --enable-builtin-atomics.

This PR *does not* enable builtin atomics by default.

@artpol84 I ran performance tests before and after these commits and see no performance regression. You should probably do the same just to be sure.

### CODE REVIEW

**Comment by hppritcha** [comment]
Is commit 19d3884 (add timer support) really related to this PR?

**Comment by hjelmn** [comment]
@hppritcha It is related to the ARMv8 stuff but I can remove it if it is a problem.

File: opal/include/opal/sys/arm64/timer.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+#include <sys/times.h>
+
+typedef uint64_t opal_timer_t;
+
+static inline opal_timer_t
+opal_sys_timer_get_cycles(void)
+{
+    opal_timer_t ret;
+
```
Comment (hjelmn) [inline opal/include/opal/sys/arm64/timer.h]: Not sure. @abjoshi-brcm What do you think? I can leave this commit out of this PR to be safe if we don't have consensus.

**Comment by hjelmn** [comment]
@hppritcha I am pulling the PGI inline asm out of this one and putting it into #2177. That way if this PR fails we still have PGI on ppc fixed.

File: opal/include/opal/sys/arm64/timer.h (hunk 1 above)
Comment (shamisp) [inline opal/include/opal/sys/arm64/timer.h]: @hjelmn - we need good timer (as part of this PR or separate one).

**Comment by bosilca** [comment]
The arm64 timer should be monotonic isn't it?

**Comment by shamisp** [comment]
correct.

File: opal/include/opal/sys/arm64/timer.h (hunk 1 above)
Comment (abjoshi-brcm) [inline opal/include/opal/sys/arm64/timer.h]: @shamisp  
You are right, reads can occur out of order and so it should be
isb(); asm volatile("mrs %0, cntvct_el0" : "=r" (ret));  
And yes, this timer increments at a fixed frequency.

Comment (shamisp) [inline opal/include/opal/sys/arm64/timer.h]: @bosilca - AFAIK this is the timer that is invoked in opal_progress (after your fix). Do we use it in some other parts of the code ?

**Comment by jsquyres** [comment]
@hjelmn Is this complete?

**Comment by hjelmn** [comment]
@jsquyres Yes. Though it sounds like there is a desire from ARM to have the timer come over as well. I can either re-add it or file a separate PR.

**Comment by shamisp** [comment]
Yep, we need the timer.

**Comment by bosilca** [comment]
This is not yet complete. Based on the discussions above there are 2 things missing the isb and the MONOTONIC define.

**Comment by hjelmn** [comment]
@shamisp Can you make the changes @bosilca mentions and PR them to master. I will add all the timer commits if that can be done by the end of the day tomorrow.

**Comment by hjelmn** [comment]
Hmm, it might be faster for me to just go ahead and make the changes on master. I will have @shamisp review.

**Comment by hjelmn** [comment]
@bosilca Looks like a MONOTONIC define isn't required as we always set it to 1. Non-monotonic timers have to undef it and redefine to 0.

**Comment by bosilca** [comment]
@hjelmn indeed thanks for pointing this out.

**Comment by shamisp** [comment]
@hjelmn can you please forward the original patch (timer) that you have been trying to merge ?

File: opal/include/opal/sys/arm64/timer.h (hunk 1 above)
Comment (shamisp) [inline opal/include/opal/sys/arm64/timer.h]: @abjoshi-brcm FYI, I did some additional checks, which confirmed that isb() is required.

**Comment by hjelmn** [comment]
@bosilca, @shamisp The ARM64 timer is back in the PR. Please review.

**Comment by shamisp** [comment]
@hjelmn 👍

**Comment by jsquyres** [comment]
@shamisp Can you give an actual github review?  That's what we're using these days -- not just the 👍

**Comment by hppritcha** [comment]
@jsquyres I think this is ready to go

**Pull request merged**